lib/media/drm_engine.js

```diff
@@ -420,27 +476,30 @@ shaka.media.DrmEngine = class {
           () => this.closeOpenSessions_());
     }
 
-    let setMediaKeys = this.video_.setMediaKeys(this.mediaKeys_);
-    setMediaKeys = setMediaKeys.catch((exception) => {
-      goog.asserts.assert(exception instanceof Error, 'Wrong error type!');
-      return Promise.reject(new shaka.util.Error(
-          shaka.util.Error.Severity.CRITICAL,
-          shaka.util.Error.Category.DRM,
-          shaka.util.Error.Code.FAILED_TO_ATTACH_TO_VIDEO,
-          exception.message));
-    });
+    const nonEmptyInitData = this.currentDrmInfo_.initData.find(
+        (initDataOverride) => initDataOverride.initData.length > 0);
 
-    await setMediaKeys;
-    this.destroyer_.ensureNotDestroyed();
+    /**
+     * We can attach media keys before the playback actually begins when:
+     *  - Using legacy implementations requires MediaKeys to be set before
+     *    having webkitneedkey / msneedkey event, which will be translated as
+     *    an encrypted event by the polyfills
+     *  - Some initData already has been generated (through the manifest)
+     *  - In case of an offline session
+     */
+    if (nonEmptyInitData ||
+        shaka.util.Platform.isMediaKeysPolyfilled() ||
+        this.offlineSessionIds_.length) {
+      await this.attachMediaKeys_();
+    }
 
     this.createOrLoad();
-    if (!this.currentDrmInfo_.initData.length &&
-        !this.offlineSessionIds_.length) {
-      // Explicit init data for any one stream or an offline session is
-      // sufficient to suppress 'encrypted' events for all streams.
-      const cb = (e) => this.newInitData(
-          e.initDataType, shaka.util.BufferUtils.toUint8(e.initData));
-      this.eventManager_.listen(this.video_, 'encrypted', cb);
+
+    // Explicit init data for any one stream or an offline session is
+    // sufficient to suppress 'encrypted' events for all streams.
+    if (!nonEmptyInitData && !this.offlineSessionIds_.length) {
```

Comment:
This variable name and this condition make a triple-negative that is hard for my brain to parse and comprehend.

If not non-empty = if empty?
If not non-empty = if not full?

I think this might be easier to read if we rename `nonEmptyInitData` to `manifestInitData`, since it comes from the manifest parser.  That would leave us with conditions like:

```js
  if (manifestInitData || isMediaKeysPolyfilled() || offlineSessionIds.length) {
    await attachMediaKeys();
  }
```

```js
  if (!manifestInitData && !offlineSessionIds.length) {
    listen(video, 'encrypted');
  }
```